### PR TITLE
Stopping old queue consumers during upgrade

### LINKED ISSUE #23540: Queue consumers
### Preconditions (*)
Then you deploy a next release you still have queue consumers running from the old releases. And where is no way to stop them

### Steps to reproduce (*)
run cron on the first release
release new version
run cron on the second release
release new version
run cron on the third release

### Expected result (*)
only new process running
```
releases/745/bin/magento queue:consumers:start product_action_attribute.website.update
```

### Actual result (*)
still running old processes
```
releases/743/bin/magento queue:consumers:start product_action_attribute.website.update
releases/744/bin/magento queue:consumers:start product_action_attribute.website.update
releases/745/bin/magento queue:consumers:start product_action_attribute.website.update
```

### PR DESCRIPTION
**Pull request by swnsma**
### Description (*)
Put poison pill after each 'setup:upgrade' execution.

### Fixed Issues (if relevant)
1. Fixes magento/magento2#23540

### Steps for testing:
1. Check "version" value in "queue_poison_pill" table in Magento instance database.
2. Upgrade Magento instance.
3. Verify "version" value in "queue_poison_pill" table has changed in Magento instance database.
4. Execute queue:consumers:restart command.
5. Verify "version" value in "queue_poison_pill" table has changed in Magento instance database.

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by hostep** [comment]
This also kind-of fixes https://github.com/magento/community-features/issues/181 and problem 2 mentioned over [here](https://github.com/magento/architecture/blob/master/design-documents/asynchronous-operations/consumer-processes-improvements.md#problem-2-deployment-problems) I think.

**Comment by swnsma** [comment]
@magento run all tests

**Comment by coderimus** [comment]
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE, Integration Tests, Magento Health Index

**Comment by coderimus** [comment]
@magento run Functional Tests B2B, Functional Tests CE, Magento Health Index

**Comment by coderimus** [comment]
Hello @swnsma 
Thank you for the fix! As for me, it is a really great solution which removes headache for everyone who uses the MQ in the Magento. I think that the approach you provided is applicable and it uses in other core modules also, as I can see. No need to have another CI command in that case 👍

I approve this PR in a comment for now but we have the Health Index failed and without assistance from the core team we are not able to move forward so let's call @sivaschenko and @sidolov for help :) Please, assist me with the next step: just approve it and move for testing or we need to do something else in case of failed Health Index with the `Difference with mainline detected` message?

**Comment by sivaschenko** [comment]
Thanks for the pull request @swnsma !

I am moving this pull request to "To Apporve" status to process internal approval.

Approval ticket: https://jira.corp.magento.com/browse/MC-40323

cc @sidolov

**Review by coderimus — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @coderimus, thank you for the review.
[ENGCOM-8607](https://jira.corp.magento.com/browse/ENGCOM-8607) has been created to process this Pull Request
:eight_spoked_asterisk: @coderimus, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Comment by ihor-sviziev** [comment]
@swnsma good job! Really good improvement! The only one thing I would like to see there additionally - test coverage. Can we cover running setup:upgrade command in kind of integration test and compare poison pill value before & after?

**Comment by coderimus** [comment]
Hi @ihor-sviziev 
Agree with your test request and we definitely need this. But in that case the PR, as I understood, will pass internal approval from the core team and if everything fine we can deliver it and create tests.

**Comment by JeroenBoersma** [comment]
I have one con on this solution, we don't want to run `setup:upgrade --keep-generated` on every release, only for those who add or update things in the database... This is checked with `setup:db:status` and `app:config:status` if these report a non-change we just do a symlink change and have near zero downtime.

So, this solution still doesn't fix this behavior because then the old runners keep running...

A simple solution would be having the possibility to also run this from commandline, then we would ran that if db and app are up-to-date

**Comment by hostep** [comment]
Ah yes good point @JeroenBoersma, a dedicated command line option would still be useful in this case next to the automatic `setup:upgrade` renewal.

**Comment by swnsma** [comment]
Hi!
I have added integration tests coverage.

@sivaschenko, do we need to include the CLI command implementation to this PR?
If yes, do we need to receive an extra approval for command implementation?

**Comment by sivaschenko** [comment]
Thank you for the test coverage @swnsma ! The approval process for this pull request is still in progress.

**Comment by hostep** [comment]
So ... Should we also create a new dedicated command for this (based on @JeroenBoersma's remark)?

**Comment by ihor-sviziev** [comment]
@hostep, from my perspective - yes, we should add it.
@sivaschenko, what do you think?

**Comment by coderimus** [comment]
@ihor-sviziev @sivaschenko @hostep @swnsma 
I think we should add a dedicated command for this to make the process more flexible for the deployment with the CLI command.

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Could you also write some test for the cli command?

File: app/code/Magento/MessageQueue/Test/Integration/PoisonPillAppyAfterCommandRunTest.php
Comment (ihor-sviziev) [inline app/code/Magento/MessageQueue/Test/Integration/PoisonPillAppyAfterCommandRunTest.php]: Can we add that before running the command that's the latest version?
In both tests

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

**Comment by ihor-sviziev** [comment]
@magento run all tests

File: app/code/Magento/MessageQueue/Test/Integration/PoisonPillAppyAfterCommandRunTest.php
Comment (swnsma) [reply app/code/Magento/MessageQueue/Test/Integration/PoisonPillAppyAfterCommandRunTest.php]: @ihor-sviziev do you mean to run "getLatestVersion" again instead of "isLatestVersion" ?

Comment (ihor-sviziev) [reply app/code/Magento/MessageQueue/Test/Integration/PoisonPillAppyAfterCommandRunTest.php]: Sorry, I just got confused. Everything looks good now

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Please double-check the integration and web API test failures

**Comment by engcom-Hotel** [comment]
@magento run Functional Tests EE, WebAPI Tests

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
Hello @shawnabramson,
I'll help with Integration Tests.

**Comment by engcom-Hotel** [comment]
@magento run Functional Tests B2B

**Comment by ihor-sviziev** [comment]
@magento run Integration Tests

**Comment by ihor-sviziev** [comment]
@engcom-Hotel still failing

**Comment by ihor-sviziev** [comment]
@magento run Integration Tests

**Comment by engcom-Hotel** [comment]
@ihor-sviziev, 
PoisonPillApplyDuringSetupUpgradeTest fix in progress.

File: app/code/Magento/MessageQueue/Console/RestartConsumerCommand.php
Comment (vasylmalanka) [inline app/code/Magento/MessageQueue/Console/RestartConsumerCommand.php]: Many DocBlocks of this PR violate Magento's standards (https://devdocs.magento.com/guides/v2.4/coding-standards/docblock-standard-general.html)

**Review by vasylmalanka — CHANGES_REQUESTED** [review]
Please change DocBlocks to follow Magento standards

File: app/code/Magento/MessageQueue/Console/RestartConsumerCommand.php
Comment (swnsma) [reply app/code/Magento/MessageQueue/Console/RestartConsumerCommand.php]: Thank you for your review!
May I ask you to be more specific?

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run Unit Tests

**Comment by engcom-Hotel** [comment]
@magento run Database Compare, Functional Tests B2B, Functional Tests CE, Functional Tests EE, Integration Tests, Magento Health Index, Sample Data Tests B2B, Sample Data Tests CE, Sample Data Tests EE, Semantic Version Checker, Static Tests, WebAPI Tests

**Comment by engcom-Hotel** [comment]
@magento run Functional Tests B2B

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests B2B

File: app/code/Magento/MessageQueue/Console/RestartConsumerCommand.php
Comment (vasylmalanka) [reply app/code/Magento/MessageQueue/Console/RestartConsumerCommand.php]: For example,

> The @inheritdoc tag SHOULD NOT be used. If a child class method requires a long description to explain its purpose, it may use @inheritdoc to indicate the new description is intended as an addition to the parent method description. In general, such method overrides are a code smell and should be used as an incentive to make the code more self-documenting if possible.
https://devdocs.magento.com/guides/v2.4/coding-standards/docblock-standard-general.html#inheritdoc

> The declaration of all arguments (if any) using the @param tag, unless the argument type is indicated in the method signature.
https://devdocs.magento.com/guides/v2.4/coding-standards/docblock-standard-general.html#functions-methods

> In general, method return type signatures should be prefered over @return type annotations.
https://devdocs.magento.com/guides/v2.4/coding-standards/docblock-standard-general.html#return

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-8607](https://jira.corp.magento.com/browse/ENGCOM-8607) has been created to process this Pull Request
:eight_spoked_asterisk: @ihor-sviziev, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Comment by engcom-Oscar** [comment]
@engcom-Foxtrot can you please check this PR, it require dev-experience.

**Comment by engcom-Foxtrot** [comment]
QA passed

**Comment by engcom-Foxtrot** [comment]
Magento Health Index failure should be approved.

**Comment by sivaschenko** [comment]
Pending approval request: https://jira.corp.magento.com/browse/MC-40323

**Comment by keharper** [comment]
@swnsma Be sure to open a PR on devdocs when this PR is merged.

**Comment by hostep** [comment]
Hi folks, what's the status here?
Would be great if this could get merged soon, so it can be included in an upcoming Magento version (for 2.4.3 it will probably already be too late now?)

**Comment by hostep** [comment]
Hi @sivaschenko: what's the status of MC-40323, why does the approval take this long? Thanks!

**Comment by swnsma** [comment]
Hi @sidolov, @sivaschenko,

Any future for this PR?

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by bartoszkubicki** [comment]
Hi, what is the status of this one? Looking forward to have it merged into codebase.

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@swnsma, could you please fix failing static tests

**Comment by ihor-sviziev** [comment]
@magento run Static Tests

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
Hi @engcom-Alfa,
I saw that you tried to move to "accept" status in https://github.com/magento/magento2/pull/31495#event-6489856047
Does it mean we got the approval for the changes?

**Comment by engcom-Alfa** [comment]
> Hi @engcom-Alfa, I saw that you tried to move to "accept" status in [#31495 (comment)](https://github.com/magento/magento2/pull/31495#event-6489856047) Does it mean we got the approval for the changes?
> 
> PS: I moved this PR back to "pending approval", as that was the last status before resolving conflict, and the changes still look good to me

Sure, will cross check and process it accordingly.
Thanks for the info :)

**Comment by engcom-Alfa** [comment]
@magento run Functional Tests B2B, Magento Health Index

**Comment by PascalBrouwers** [comment]
@sidolov @sivaschenko can we get this merged?

**Comment by ihor-sviziev** [comment]
also pinging @ishakhsuvarov

**Comment by engcom-Hotel** [comment]
Raised new internal approval request JIRA [ticket](https://jira.corp.adobe.com/browse/AC-6932) on the adobe portal for Health Index test failure. Waiting for the approvals.

**Comment by abhattGlo** [comment]
@magento run Magento Health Index

**Comment by ishakhsuvarov** [comment]
@ihor-sviziev I am working to push this ticket forward

**Review by ishakhsuvarov — APPROVED** [review]

**Comment by ishakhsuvarov** [comment]
@swnsma @ihor-sviziev This actually got merged, however last commit with 2.4 sync did not get included in the internal merge, thus creating this confusiong situation.

Not sure why github thinks it's "closed" and not "merged" but there is 0 files and commits in the changeset.

Sorry for the confusion.

**Comment by swnsma** [comment]
@ishakhsuvarov, thank you!
After so many months this seems to be Christmas Miracle! :)

**Comment by fritzmg** [comment]
Since the exact commits are lost (or may be someone has a link?) - was this actually already released, or will this be in the next release?

**Comment by hostep** [comment]
Yes, it will be included in Magento 2.4.6 (it's included in 2.4.6-beta1 at least)

And I've found these commits: https://github.com/magento/magento2/search?q=%22magento%2Fmagento2%2323540%22&type=commits

**Pull request closed without merging**
